### PR TITLE
feat(repo): Added pre-commit hooks, removed examples, cleanup

### PR DESCRIPTION
**Pull request by rwlockbg**
1. removed unused k8s with mongodb operator - not part of this repo
2. removed examples - irrelevant to our protocol
3. fixed errors in tests (btc-server)
4. added pre-commit hooks
5. added/extended fmt and lint commands
6. added npm tools for thorough linting and formatting of tomls, jsons, etc.
7. run cargo machete
8. removed unused deps, variables etc.
9. fixed all clippy warnings and errors
10. surpressed protobuf autogenerated files

### CODE REVIEW

**Comment by birozuru** [comment]
# eBurger report

**Comment by birozuru** [comment]
# Slither report
**THIS CHECKLIST IS NOT COMPLETE**. Use `--show-ignored-findings` to show all the results.
Summary
 - [arbitrary-send-eth](#arbitrary-send-eth) (1 results) (High)
 - [missing-zero-check](#missing-zero-check) (1 results) (Low)
 - [solc-version](#solc-version) (1 results) (Informational)
 - [reentrancy-unlimited-gas](#reentrancy-unlimited-gas) (1 results) (Informational)
## arbitrary-send-eth
Impact: High
Confidence: Medium
 - [ ] ID-0
[Minting.mint(address,uint256,uint32,bytes,address)](https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L45-L84) sends eth to arbitrary user
	Dangerous calls:
	- [address(destination).transfer(amount)](https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L80)
	- [address(refundAddress).transfer(txCost)](https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L81)

https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L45-L84


## missing-zero-check
Impact: Low
Confidence: Medium
 - [ ] ID-1
[Minting.mint(address,uint256,uint32,bytes,address).refundAddress](https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L50) lacks a zero-check on :
		- [address(refundAddress).transfer(txCost)](https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L81)

https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L50


## solc-version
Impact: Informational
Confidence: High
 - [ ] ID-2
Version constraint ^0.8.13 contains known severe issues (https://solidity.readthedocs.io/en/latest/bugs.html)
	- VerbatimInvalidDeduplication
	- FullInlinerNonExpressionSplitArgumentEvaluationOrder
	- MissingSideEffectsOnSelectorAccess
	- StorageWriteRemovalBeforeConditionalTermination
	- AbiReencodingHeadOverflowWithStaticArrayCleanup
	- DirtyBytesArrayToStorage
	- InlineAssemblyMemorySideEffects
	- DataLocationChangeInInternalOverride
	- NestedCalldataArrayAbiReencodingSizeValidation.
It is used by:
	- [^0.8.13](https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L2)

https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L2


## reentrancy-unlimited-gas
Impact: Informational
Confidence: Medium
 - [ ] ID-3
Reentrancy in [Minting.mint(address,uint256,uint32,bytes,address)](https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L45-L84):
	External calls:
	- [address(destination).transfer(amount)](https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L80)
	- [address(refundAddress).transfer(txCost)](https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L81)
	Event emitted after the call(s):
	- [Mint(destination,amount,bitcoinBlockHeight,metadata)](https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L83)

https://github.com/botanix-labs/Macbeth/blob/d0840ff280b5e8fe7c1c196dea04083a396eacf5/contracts/src/Minting.sol#L45-L84

**Comment by arminsabouri** [comment]
@rwlockbg this Pr is difficult to review. I can review commit by commit but then there are commits that seemingly do the same thing (as far as the message goes) and I think these can be squashd together. For example there are 5 commits that fix cargo tomls. Can these be squashed together?

**Comment by arminsabouri** [comment]
> @rwlockbg this Pr is difficult to review. I can review commit by commit but then there are commits that seemingly do the same thing (as far as the message goes) and I think these can be squashd together. For example there are 5 commits that fix cargo tomls. Can these be squashed together?

@rwlockbg following up here. The best course here is probably to have seperate commits: 
* the ones that create the new linting rules 
* the ones that apply them on the whole codebase.

**Comment by arminsabouri** [comment]
One more point of consideration.
@PhoenixCodeCraft will this type of change impact the audit(s)? After this change is made the diff (from our main to upstream main) will be substantial. This will obscure which code changes are cosmetic vs functional changes.

**Comment by rwlockbg** [comment]
> One more point of consideration. @PhoenixCodeCraft will this type of change impact the audit(s)? After this change is made the diff (from our main to upstream main) will be substantial. This will obscure which code changes are cosmetic vs functional changes.

Just to mention here that rust files formatting has not changed at all. The changes in the rust files are all clippy - related etc.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#         language: rust
+#         pass_filenames: false
+
+#   - repo: https://github.com/lyz-code/yamlfix/
+#     rev: 1.17.0
+#     hooks:
+#       - id: yamlfix
+#         args:
+#           - -c
+#           - .yamlfix.toml
```
Comment (arminsabouri) [inline .pre-commit-config.yaml]: do we need to keep any of these comments in?

**Comment by PhoenixCodeCraft** [comment]
It shouldn't impact the quote amount, but indeed good to know and communicate so they expect it when they get the final diff to work on.

File: bin/btc-server/src/merkle.rs [hunk 2]
```diff
@@ -47,9 +47,9 @@ pub fn get_wallet_state_commitment(db: &Db) -> Result<WalletState, Error> {
     let tracked_tx_root = get_tracked_tx_set_merkle_root(db)?;
     let pending_pegouts_root = get_pending_pegouts_set_merkle_root(db)?;
     let mut engine = sha256::Hash::engine();
-    engine.write(&utxo_root).expect("to write 32 byte value");
-    engine.write(&tracked_tx_root).expect("to write 32 byte value");
-    engine.write(&pending_pegouts_root).expect("to write 32 byte value");
+    let _ = engine.write(&utxo_root).expect("to write 32 byte value");
```
Comment (arminsabouri) [inline bin/btc-server/src/merkle.rs:50]: Oops I should be erroring here. We'll make a ticket for this

File: .github/workflows/lint.yml [hunk 3]
```diff
@@ -189,12 +195,66 @@ jobs:
...
+
+      - uses: pnpm/action-setup@v4
+        name: Install pnpm
+        with:
+          version: 9
+          run_install: true
+
+      # - uses: taiki-e/install-action@cargo-machete
+      # - name: Check unused dependencies
+      #   run: cargo machete --skip-target-dir --with-metadata
```
Comment (arminsabouri) [inline .github/workflows/lint.yml]: can we remove this?

**Review by arminsabouri — COMMENTED** [review]
Overall this looks great. Im not entirely sure what the benefit of having precommit hook checks run on CI are. My impression of pre-commit checks are that they ensure the CI checks will *probably* succeed. 

Other than that I just had a few comments and questions

File: .pre-commit-config.yaml (hunk 1 above)
Comment (rwlockbg) [reply .pre-commit-config.yaml]: These are no longer comments , can you pls double-check. You might have an older branch or so?

File: bin/btc-server/src/merkle.rs (hunk 2 above)
Comment (rwlockbg) [reply bin/btc-server/src/merkle.rs:50]: You can def do so, I made it so it passes the clippy checks now.

File: .github/workflows/lint.yml (hunk 3 above)
Comment (rwlockbg) [reply .github/workflows/lint.yml]: done

**Comment by arminsabouri** [comment]
@birozuru to take a look at the pre-commit job

**Comment by arminsabouri** [comment]
Pre-commit job seems to run for multiple hours then fails. Do we need this job? A pre-commit should be something we run locally before we commit to ensure CI will _probably_ pass.

File: crates/net/network/src/frost/mod.rs [hunk 4]
```diff
@@ -15,19 +15,22 @@ pub mod messages;
...
-}
+// impl ProtocolState {
+//     /// Constructs a new Protocol State.
+//     pub const fn new(
+//         events: mpsc::UnboundedSender<FrostProtocolEvent>,
+//         my_peer_id: PeerId,
+//     ) -> Self {
+//         Self { events, my_peer_id }
+//     }
+// }
```
Comment (arminsabouri) [inline crates/net/network/src/frost/mod.rs]: delete this instead of comment

Comment (rwlockbg) [reply crates/net/network/src/frost/mod.rs]: removed

**Pull request merged**